sql/2020/22_HTTP_2/number_of_http_sites_returning_upgrade_http_header_containing_h2.sql

```diff
@@ -0,0 +1,30 @@
+# standardSQL
+# Number of HTTP (not HTTPS) sites which return upgrade HTTP header containing h2.
+CREATE TEMPORARY FUNCTION getUpgradeHeader(payload STRING)
+RETURNS STRING
+LANGUAGE js AS """
+  try {
+    var $ = JSON.parse(payload);
+    var headers = $.response.headers;
+    var st = headers.find(function(e) { 
+      return e['name'].toLowerCase() === 'upgrade'
+    });
+    return st['value'];
+  } catch (e) {
+    return '';
+  }
+""";
+
+SELECT 
+  client,
+  firstHtml,  
+  COUNT(0) AS num_requests
+FROM 
+  `httparchive.almanac.requests`
+WHERE
+  date='2020-08-01' AND 
+  url LIKE "http://%" AND
+  getUpgradeHeader(payload) LIKE "%h2%"
```

Comment:
Consider moving this out of the `WHERE` clause, and changing `num_requests` to a conditional count:

```sql
COUNTIF(getUpgradeHeader(payload) LIKE "%h2%") AS num_requests
```

That way you can use `COUNT(0) AS total` and also calculate the percent of HTTP sites that can upgrade to H2.

Reply:
@rviscomi @bazzadp I made the following changes:

Added the `COUNTIF(getUpgradeHeader(payload) LIKE "%h2%") AS num_requests` and `Count(0) as total` to the `SELECT`

Added `JSON_EXTRACT_SCALAR(payload, '$._protocol') as http_version` to `SELECT` and `GROUP BY`

Added `JSON_EXTRACT_SCALAR(payload, '$._protocol') != ''` to `WHERE` statement to filter out any null values for `http_version` I think it makes sense to filter out nulls....let me know if you disagree

Reply:
@bazzadp Need some clarification on your comment `Look at all upgrade except “Webprotocol” to include any HTTP/3 type upgrades (which shouldn’t really be sent as should use Alt-Svc but still interested to see).`

Reply:
I don't see the protocol changes you mentioned @gregorywolf . You can ignore my comments. Let's leave looking at `h2` only. But yes lets add the protocol.